plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/LibraryClasspathContainerResolverService.java

```diff
@@ -0,0 +1,315 @@
+    return newLibraryEntry;
+  }
+
+  private SourceAttacherJob createSourceAttacherJob(IJavaProject javaProject,
+                                                    IPath containerPath,
+                                                    final LibraryFile libraryFile,
+                                                    final IProgressMonitor monitor,
+                                                    final Artifact artifact,
+                                                    IPath libraryPath) {
+    return new SourceAttacherJob(javaProject, containerPath, libraryPath, new Callable<IPath>() {
+      @Override
+      public IPath call() throws Exception {
+          return repositoryService.resolveSourceArtifact(libraryFile, artifact.getVersion(),
+                                                         monitor);
+      }
+    });
+  }
+
+  private IClasspathEntry resolveLibraryFileAttachSourceSync(final LibraryFile libraryFile)
+                                                                              throws CoreException {
+    final Artifact artifact = repositoryService.resolveArtifact(libraryFile, new NullProgressMonitor());
+    IPath libraryPath = new Path(artifact.getFile().getAbsolutePath());
+    IPath sourceAttachmentPath = null;
+    sourceAttachmentPath = repositoryService.resolveSourceArtifact(libraryFile,
+                                                                   artifact.getVersion(),
+                                                                   new NullProgressMonitor());
+    final IClasspathEntry newLibraryEntry =
+        JavaCore.newLibraryEntry(libraryPath,
+                                 sourceAttachmentPath,
+                                 null /*  sourceAttachmentRootPath */,
+                                 getAccessRules(libraryFile.getFilters()),
+                                 getClasspathAttributes(libraryFile, artifact),
+                                 true /* isExported */);
+    return newLibraryEntry;
+
+  }
+
+  private static int getTotalwork(IClasspathEntry[] rawClasspath) {
+    int sum = 0;
+    for (IClasspathEntry element : rawClasspath) {
+      if (isLibraryClasspathEntry(element.getPath())) {
+        ++sum;
+      }
+    }
+    return sum;
+  }
+
+  private static boolean isLibraryClasspathEntry(IPath path) {
+    return path != null
+        && path.segmentCount() == 2
+        && Library.CONTAINER_PATH_PREFIX.equals(path.segment(0));
+  }
+
+  private static String getLibraryDescription(Library library) {
+    if (!Strings.isNullOrEmpty(library.getName())) {
+      return library.getName();
+    } else {
+      return library.getId();
+    }
+  }
+
+  @Activate
+  protected void initialize() {
+    libraryFactory = new LibraryFactory();
+    serializer = new LibraryClasspathContainerSerializer();
+    IConfigurationElement[] configurationElements =
+    RegistryFactory.getRegistry().getConfigurationElementsFor(LIBRARIES_EXTENSION_POINT);
```

Comment:
Since you're configuring this as an OSGi declarative service you can have the `IExtensionRegistry` be provided to you — it's available as an OSGi service too.